app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,34 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+
+  def facebook
+    auth = request.env['omniauth.auth']
+
+    if auth.info.email.blank?
+      redirect_to user_omniauth_authorize_path(
+        :facebook,
+        auth_type: 'rerequest',
+        scope: 'email,public_profile'
+      )
+    else
+      user = User.find_by(provider: auth.provider, uid: auth.uid)
+
+      if user.nil?
+        user = User.find_by(email: auth.info.email)
+
+        if user.nil?
+          user = User.create_from_omniauth(auth)
+        elsif user.uid != auth.uid
+          user.update_from_omniauth(auth)
+        end
+      end
```

Comment:
One other thing to note here is that I ripped out the case for when a user is signed in – just as an initial stab at keeping it cleaner / simpler for now. The primary reason for that is that I'd want to sketch out those edge cases a bit more. For example what happens in the following:

- User is signed in, tries to connect with FB account, but that FB account (e.g. uid) is connected to a different user account – we shouldn't call `user.update_with_oauth` then, since we'd try to add the same uid to two accounts.

- User is signed in and we find another user account (e.g. not `current_user`) that has the same email as the FB login they tried to use (think that's what you outlined here: https://github.com/Sevenzo/challenge-platform/blob/cbdc18a2660232ae9f8eb064b4a74dae21c66219/app/controllers/omniauth_callbacks_controller.rb#L16), don't think we could just call `update_with_oauth` either as we effectively have a different account with their FB email as we do with their `current_user` and that may cause some weirdness down the line.

In general, those cases are slightly harder to map out, and mostly wanted to make sure we took care of this initial simpler cases first (and add tests to handle all of this as well...). We don't need to worry too much about the feature spec, the request spec should be sufficient that you had started.

Not sure what the exact scope of this feature was, but I think providing a way for only non-signed in users to have the FB flow is sufficient to complete it at this stage as we think more on those other items when you're signed in (keep in mind we'd need to add an interface for that flow which we don't have at the moment either).